File: python/journal/directory_reader.py
# Directory reader - iterates across multiple journal files.

import os
from .reader import FileReader
from .compress import is_journal_file_name


class DirectoryReader:
    def __init__(self, path, readers=None):
        self._path = path
        self._readers = readers or []
        self._index = -1
        self._realtime_seek = None
        self._realtime_seek_bound = None
        self._candidates = [None] * len(self._readers)
        self._current_key = None
        self._direction = None
        self._boot_newest = _build_boot_newest(self._readers)

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc, tb):
        try:
            self.close()
        except Exception:
            if exc_type is None:
                raise
        return False

    @staticmethod
    def open(path):
        if not os.path.isdir(path):
            raise ValueError(f'not a directory: {path}')

        readers = []
        for journal_path in _collect_journal_files(path):
            try:
                reader = FileReader.open(journal_path)
            except Exception:
                reader = None
            if reader is not None:
                readers.append(reader)

        return DirectoryReader.from_readers(path, readers, allow_empty=True)

    @staticmethod
    def open_files(paths):
        readers = []
        for path in paths:
            if not is_journal_file_name(os.path.basename(path)):
                raise ValueError(f'not a journal file: {path}')
            readers.append(FileReader.open(path))
        return DirectoryReader.from_readers('<files>', readers, allow_empty=False)

    @staticmethod
    def from_readers(path, readers, allow_empty=False):
        if not readers and not allow_empty:
            raise ValueError(f'no readable journal files in {path}')
        readers.sort(key=lambda r: (r.header()['head_entry_realtime'], r.header()['head_entry_seqnum']))
        return DirectoryReader(path, readers)

    def seek_head(self):
        self._realtime_seek = None
        self._realtime_seek_bound = None
        self._index = -1
        self._current_key = None
        self._direction = None
        self._reset_candidates()
        for reader in self._readers:
            reader.seek_head()

    def seek_tail(self):
        self._realtime_seek = None
        self._realtime_seek_bound = None
        self._index = -1
        self._current_key = None
        self._direction = None
        self._reset_candidates()
        for reader in self._readers:
            reader.seek_tail()

    def seek_realtime_usec(self, usec):
        self._realtime_seek = int(usec)
        self._realtime_seek_bound = None
        self._index = -1
        self._current_key = None
        self._direction = None
        self._reset_candidates()

    def step(self):
        return self._step_merged(0)

    def step_back(self):
        return self._step_merged(1)

    def _step_merged(self, direction):
        self._prepare_merge_direction(direction)

        best = None
        for idx in range(len(self._readers)):
            self._fill_candidate(idx, direction)
            candidate = self._candidates[idx]
            if candidate is None:
                continue
            if best is None:
                best = candidate
                continue
            cmp = self._compare_entry_keys(candidate['key'], best['key'])
            if (direction == 0 and cmp < 0) or (direction == 1 and cmp > 0):
                best = candidate

        if best is None:
            self._index = -1
            self._realtime_seek_bound = None
            return False

        self._index = best['reader_index']
        self._current_key = best['key']
        self._candidates[best['reader_index']] = None
        self._realtime_seek_bound = None
        return True

    def _prepare_merge_direction(self, direction):
        if self._realtime_seek is not None:
            usec = self._realtime_seek
            self._realtime_seek = None
            for reader in self._readers:
                reader.seek_realtime_usec(usec)
            self._reset_candidates()
            self._realtime_seek_bound = (usec, direction)
            self._direction = direction
            return

        if self._direction == direction:
            return

        if self._current_key is not None:
            for reader in self._readers:
                reader.seek_realtime_usec(self._current_key['realtime'])
        elif direction == 0:
            for reader in self._readers:
                reader.seek_head()
        else:
            for reader in self._readers:
                reader.seek_tail()

        self._reset_candidates()
        self._direction = direction

    def _fill_candidate(self, reader_index, direction):
        if self._candidates[reader_index] is not None:
            return
        reader = self._readers[reader_index]

        while True:
            ok = reader.step() if direction == 0 else reader.step_back()
            if not ok:
                return
            key = reader.current_entry_key()
            if key is None:
                continue
            if self._realtime_seek_bound is not None:
                usec, seek_direction = self._realtime_seek_bound
                if (seek_direction == 0 and key['realtime'] < usec) or (
                    seek_direction == 1 and key['realtime'] > usec
                ):
                    continue
            if self._current_key is not None:
                cmp = self._compare_entry_keys(key, self._current_key)
                if (direction == 0 and cmp <= 0) or (direction == 1 and cmp >= 0):
                    continue

            self._candidates[reader_index] = {'reader_index': reader_index, 'key': key}
            return

    def _compare_entry_keys(self, a, b):
        if (
            a['boot_id'] == b['boot_id'] and
            a['monotonic'] == b['monotonic'] and
            a['realtime'] == b['realtime'] and
            a['xor_hash'] == b['xor_hash'] and
            a['seqnum_id'] == b['seqnum_id'] and
            a['seqnum'] == b['seqnum']
        ):
            return 0

        if a['seqnum_id'] == b['seqnum_id']:
            cmp = _cmp_int(a['seqnum'], b['seqnum'])
            if cmp != 0:
                return cmp

        if a['boot_id'] == b['boot_id']:
            cmp = _cmp_int(a['monotonic'], b['monotonic'])
            if cmp != 0:
                return cmp
        else:
            cmp = self._compare_boot_ids(a['boot_id'], b['boot_id'])
            if cmp != 0:
                return cmp

        cmp = _cmp_int(a['realtime'], b['realtime'])
        if cmp != 0:
            return cmp
        return _cmp_int(a['xor_hash'], b['xor_hash'])

    def _compare_boot_ids(self, a, b):
        a_newest = self._boot_newest.get(a)
        b_newest = self._boot_newest.get(b)
        if a_newest is None or b_newest is None or a_newest['machine_id'] != b_newest['machine_id']:
            return 0
        return _cmp_int(a_newest['realtime'], b_newest['realtime'])

    def _reset_candidates(self):
        self._candidates = [None] * len(self._readers)

    def next(self):
        return self.step()

    def previous(self):
        return self.step_back()

    def get_entry(self):
        if self._index < 0 or self._index >= len(self._readers):
            return None
        return self._readers[self._index].get_entry()

    def get_realtime_usec(self):
        if self._index < 0 or self._index >= len(self._readers):
            return 0
        return self._readers[self._index].get_realtime_usec()

    def get_cursor(self):
        if self._index < 0 or self._index >= len(self._readers):
            return None
        return self._readers[self._index].get_cursor()

    def test_cursor(self, cursor):
        if self._index < 0 or self._index >= len(self._readers):
            return False
        return self._readers[self._index].test_cursor(cursor)

    def add_match(self, data):
        for r in self._readers:
            r.add_match(data)
        self._reset_merge_state()

    def add_disjunction(self):
        for r in self._readers:
            r.add_disjunction()
        self._reset_merge_state()

    def add_conjunction(self):
        for r in self._readers:
            r.add_conjunction()
        self._reset_merge_state()

    def flush_matches(self):
        for r in self._readers:
            r.flush_matches()
        self._reset_merge_state()

    def enumerate_fields(self):
        fields = set()
        for r in self._readers:
            fields.update(r.enumerate_fields())
        return sorted(fields)

    def query_unique(self, field_name):
        seen = set()
        results = []
        for r in self._readers:
            vals = r.query_unique(field_name)
            for v in vals:
                key = v.hex()
                if key not in seen:
                    seen.add(key)
                    results.append(v)
        return results

    def current_entry_key(self):
        if self._index < 0 or self._index >= len(self._readers):
            return None
        return self._readers[self._index].current_entry_key()

    def visit_entry_payloads(self, visitor):
        if self._index < 0 or self._index >= len(self._readers):
            raise ValueError('no entry at current position')
        return self._readers[self._index].visit_entry_payloads(visitor)

    def collect_entry_payloads(self):
        if self._index < 0 or self._index >= len(self._readers):
            raise ValueError('no entry at current position')
        return self._readers[self._index].collect_entry_payloads()

    def get_entry_payload(self, field_name):
        if self._index < 0 or self._index >= len(self._readers):
            return None
        return self._readers[self._index].get_entry_payload(field_name)

    def get_raw(self, field_name):
        if self._index < 0 or self._index >= len(self._readers):
            return None
        return self._readers[self._index].get_raw(field_name)

    def get_raw_values(self, field_name):
        if self._index < 0 or self._index >= len(self._readers):
            return []
        return self._readers[self._index].get_raw_values(field_name)

    def entry_data_restart(self):
        if self._index < 0 or self._index >= len(self._readers):
            raise ValueError('no entry at current position')
        return self._readers[self._index].entry_data_restart()

    def enumerate_entry_payload(self):
        if self._index < 0 or self._index >= len(self._readers):
            return None
        return self._readers[self._index].enumerate_entry_payload()

    def clear_entry_data_state(self):
        if self._index < 0 or self._index >= len(self._readers):
            return
        self._readers[self._index].clear_entry_data_state()

    def list_boots(self):
        boots = {}
        for r in self._readers:
            h = r.header()
            boot_id = h['tail_entry_boot_id'].hex()
            first = h['head_entry_realtime']
            last = h['tail_entry_realtime']
            if boot_id in boots:
                boots[boot_id] = (min(boots[boot_id][0], first), max(boots[boot_id][1], last))
            else:
                boots[boot_id] = (first, last)

        result = []
        for boot_id, (first_entry, last_entry) in boots.items():
            result.append({
                'index': 0,
                'boot_id': boot_id,
                'first_entry': first_entry,
                'last_entry': last_entry,
            })
        result.sort(key=lambda x: x['first_entry'])
        base = 1 - len(result)
        for i, b in enumerate(result):
            b['index'] = base + i
        return result

    def close(self):
        close_err = None
        for r in self._readers:
            try:
                r.close()
            except Exception as e:
                if close_err is None:
                    close_err = e
        if close_err is not None:
            raise close_err

    def _reset_merge_state(self):
        self._index = -1
        self._current_key = None
        self._direction = None
        self._realtime_seek_bound = None
        self._reset_candidates()


def _collect_journal_files(path):
    files = []
    with os.scandir(path) as scan:
        entries = list(scan)
    for entry in entries:
        if _entry_is_file(entry) and is_journal_file_name(os.path.basename(entry.name)):
            files.append(entry.path)
    for entry in entries:
        if not _is_journal_subdir_name(entry.name) or not _entry_is_dir(entry):
            continue
        try:
            with os.scandir(entry.path) as children:
                for child in children:
                    if _entry_is_file(child) and is_journal_file_name(os.path.basename(child.name)):
                        files.append(child.path)
        except OSError:
            continue
    return sorted(files)


def _entry_is_file(entry):
    try:
        return entry.is_file()
    except OSError:
        return False


def _entry_is_dir(entry):
    try:
        return entry.is_dir()
    except OSError:
        return False


def _is_journal_subdir_name(name):
    if '.' in name:
        return False
    return _id128_string_valid(name)


def _id128_string_valid(value):
    if len(value) == 32:
        return all(ch in '0123456789abcdefABCDEF' for ch in value)
    if len(value) == 36:
        for idx, ch in enumerate(value):
            if idx in (8, 13, 18, 23):
                if ch != '-':
                    return False
            elif ch not in '0123456789abcdefABCDEF':
                return False
        return True
    return False


def _build_boot_newest(readers):
    newest = {}
    for reader in readers:
        header = reader.header()
        boot_id = header['tail_entry_boot_id']
        if boot_id == b'\x00' * 16:
            continue
        current = newest.get(boot_id)
        if current is None or header['tail_entry_monotonic'] > current['monotonic']:
            newest[boot_id] = {
                'machine_id': header['machine_id'],
                'monotonic': header['tail_entry_monotonic'],
                'realtime': header['tail_entry_realtime'],
            }
    return newest


def _cmp_int(a, b):
    return (a > b) - (a < b)

File: node/src/lib/directory-reader.js
// Directory reader for iterating across multiple journal files.

import { readdirSync, statSync } from 'node:fs';
import { join } from 'node:path';
import { isJournalFileName } from './compress.js';
import { FileReader, FilterBuilder } from './reader.js';
import { uuidToString } from './binary.js';

export class DirectoryReader {
  constructor() {
    this.readers = [];
    this.index = -1;
    this.filter = null;
    this.realtimeSeek = null;
    this.realtimeSeekBound = null;
    this.candidates = [];
    this.currentKey = null;
    this.direction = null;
    this.bootNewest = new Map();
  }

  static open(path) {
    const dr = new DirectoryReader();

    for (const file of collectJournalFiles(path)) {
      try {
        dr.readers.push(FileReader.open(file));
      } catch { /* skip unreadable */ }
    }

    return DirectoryReader.fromReaders(dr.readers, true);
  }

  static openFiles(paths) {
    const readers = [];
    for (const path of paths) {
      if (!isJournalFileName(String(path).split('/').pop())) {
        throw new Error(`not a journal file: ${path}`);
      }
      readers.push(FileReader.open(path));
    }
    return DirectoryReader.fromReaders(readers, false);
  }

  static fromReaders(readers, allowEmpty = false) {
    if (readers.length === 0 && !allowEmpty) throw new Error('no journal files found');
    const dr = new DirectoryReader();
    dr.readers = readers;
    dr.candidates = Array(readers.length).fill(null);
    dr.bootNewest = buildBootNewest(readers);
    dr.readers.sort((a, b) => {
      const dt = a.header.head_entry_realtime - b.header.head_entry_realtime;
      if (dt !== 0n) return dt > 0n ? 1 : -1;
      const ds = a.header.head_entry_seqnum - b.header.head_entry_seqnum;
      return ds > 0n ? 1 : ds < 0n ? -1 : 0;
    });
    return dr;
  }

  close() {
    let firstError = null;
    for (const r of this.readers) {
      try {
        r.close();
      } catch (error) {
        if (!firstError) firstError = error;
      }
    }
    this.readers = [];
    if (firstError) throw firstError;
  }

  seekHead() {
    this.realtimeSeek = null;
    this.realtimeSeekBound = null;
    this.index = -1;
    this.currentKey = null;
    this.direction = null;
    this._resetCandidates();
    for (const reader of this.readers) reader.seekHead();
  }

  seekTail() {
    this.realtimeSeek = null;
    this.realtimeSeekBound = null;
    this.index = -1;
    this.currentKey = null;
    this.direction = null;
    this._resetCandidates();
    for (const reader of this.readers) reader.seekTail();
  }

  seekRealtimeUsec(usec) {
    this.realtimeSeek = BigInt(usec);
    this.realtimeSeekBound = null;
    this.index = -1;
    this.currentKey = null;
    this.direction = null;
    this._resetCandidates();
  }

  step() {
    return this._stepMerged(0, true);
  }

  stepBack() {
    return this._stepMerged(1, true);
  }

  next() {
    return this.step();
  }

  previous() {
    return this.stepBack();
  }

  _stepMerged(direction, applyFilter) {
    this._prepareMergeDirection(direction);

    let best = null;
    for (let i = 0; i < this.readers.length; i++) {
      this._fillCandidate(i, direction, applyFilter);
      const candidate = this.candidates[i];
      if (!candidate) continue;
      if (!best) {
        best = candidate;
        continue;
      }
      const cmp = this._compareEntryKeys(candidate.key, best.key);
      if ((direction === 0 && cmp < 0) || (direction === 1 && cmp > 0)) {
        best = candidate;
      }
    }

    if (!best) {
      this.index = -1;
      this.realtimeSeekBound = null;
      return false;
    }

    this.index = best.readerIndex;
    this.currentKey = best.key;
    this.candidates[best.readerIndex] = null;
    this.realtimeSeekBound = null;
    return true;
  }

  _prepareMergeDirection(direction) {
    if (this.realtimeSeek !== null) {
      const usec = this.realtimeSeek;
      this.realtimeSeek = null;
      for (const reader of this.readers) reader.seekRealtimeUsec(usec);
      this._resetCandidates();
      this.realtimeSeekBound = { usec, direction };
      this.direction = direction;
      return;
    }

    if (this.direction === direction) return;

    if (this.currentKey) {
      for (const reader of this.readers) reader.seekRealtimeUsec(this.currentKey.realtime);
    } else if (direction === 0) {
      for (const reader of this.readers) reader.seekHead();
    } else {
      for (const reader of this.readers) reader.seekTail();
    }

    this._resetCandidates();
    this.direction = direction;
  }

  _fillCandidate(readerIndex, direction, applyFilter) {
    if (this.candidates[readerIndex]) return;
    const reader = this.readers[readerIndex];

    for (;;) {
      const ok = direction === 0 ? reader.next() : reader.previous();
      if (!ok) return;

      let key = null;
      try {
        key = reader.currentEntryKey();
      } catch {
        continue;
      }
      if (!key) continue;
      if (applyFilter && this.filter) {
        let entry = null;
        try {
          entry = reader.getEntry();
        } catch {
          continue;
        }
        if (!entry || !this.filter.matches(entry)) continue;
      }

      if (this.realtimeSeekBound) {
        const { usec, direction: seekDirection } = this.realtimeSeekBound;
        if ((seekDirection === 0 && key.realtime < usec) || (seekDirection === 1 && key.realtime > usec)) {
          continue;
        }
      }
      if (this.currentKey) {
        const cmp = this._compareEntryKeys(key, this.currentKey);
        if ((direction === 0 && cmp <= 0) || (direction === 1 && cmp >= 0)) continue;
      }

      this.candidates[readerIndex] = { readerIndex, key };
      return;
    }
  }

  _entryKey(reader, entry) {
    return {
      seqnumId: Buffer.from(reader.header.seqnum_id),
      seqnum: entry.seqnum,
      bootId: Buffer.from(entry.boot_id),
      monotonic: entry.monotonic,
      realtime: entry.realtime,
      xorHash: entry.xor_hash,
    };
  }

  _compareEntryKeys(a, b) {
    if (bufferEqual(a.bootId, b.bootId) &&
        a.monotonic === b.monotonic &&
        a.realtime === b.realtime &&
        a.xorHash === b.xorHash &&
        bufferEqual(a.seqnumId, b.seqnumId) &&
        a.seqnum === b.seqnum) {
      return 0;
    }
    if (bufferEqual(a.seqnumId, b.seqnumId)) {
      const cmp = cmpBigInt(a.seqnum, b.seqnum);
      if (cmp !== 0) return cmp;
    }
    if (bufferEqual(a.bootId, b.bootId)) {
      const cmp = cmpBigInt(a.monotonic, b.monotonic);
      if (cmp !== 0) return cmp;
    } else {
      const cmp = this._compareBootIds(a.bootId, b.bootId);
      if (cmp !== 0) return cmp;
    }
    const realtimeCmp = cmpBigInt(a.realtime, b.realtime);
    if (realtimeCmp !== 0) return realtimeCmp;
    return cmpBigInt(a.xorHash, b.xorHash);
  }

  _compareBootIds(a, b) {
    const aNewest = this.bootNewest.get(bufferKey(a));
    const bNewest = this.bootNewest.get(bufferKey(b));
    if (!aNewest || !bNewest || !bufferEqual(aNewest.machineId, bNewest.machineId)) return 0;
    return cmpBigInt(aNewest.realtime, bNewest.realtime);
  }

  _resetCandidates() {
    this.candidates = Array(this.readers.length).fill(null);
  }

  getEntry() {
    if (this.index < 0 || this.index >= this.readers.length) return null;
    return this.readers[this.index].getEntry();
  }

  getCursor() {
    if (this.index < 0 || this.index >= this.readers.length) return null;
    return this.readers[this.index].getCursor();
  }

  testCursor(cursor) { return this.getCursor() === cursor; }

  getRealtimeUsec() {
    if (this.index < 0 || this.index >= this.readers.length) return 0n;
    return this.readers[this.index].getRealtimeUsec();
  }

  currentEntryKey() {
    if (this.index < 0 || this.index >= this.readers.length) return null;
    return this.readers[this.index].currentEntryKey();
  }

  visitEntryPayloads(visitor) {
    if (this.index < 0 || this.index >= this.readers.length) throw new Error('no entry at current position');
    return this.readers[this.index].visitEntryPayloads(visitor);
  }

  collectEntryPayloads() {
    if (this.index < 0 || this.index >= this.readers.length) throw new Error('no entry at current position');
    return this.readers[this.index].collectEntryPayloads();
  }

  getEntryPayload(fieldName) {
    if (this.index < 0 || this.index >= this.readers.length) return null;
    return this.readers[this.index].getEntryPayload(fieldName);
  }

  getRaw(fieldName) {
    if (this.index < 0 || this.index >= this.readers.length) return null;
    return this.readers[this.index].getRaw(fieldName);
  }

  getRawValues(fieldName) {
    if (this.index < 0 || this.index >= this.readers.length) return [];
    return this.readers[this.index].getRawValues(fieldName);
  }

  entryDataRestart() {
    if (this.index < 0 || this.index >= this.readers.length) throw new Error('no entry at current position');
    return this.readers[this.index].entryDataRestart();
  }

  enumerateEntryPayload() {
    if (this.index < 0 || this.index >= this.readers.length) return null;
    return this.readers[this.index].enumerateEntryPayload();
  }

  clearEntryDataState() {
    if (this.index < 0 || this.index >= this.readers.length) return;
    this.readers[this.index].clearEntryDataState();
  }

  addMatch(data) {
    if (!this.filter) this.filter = new FilterBuilder();
    this.filter.addMatch(data);
    this._resetMergeState();
  }

  addDisjunction() {
    if (!this.filter) this.filter = new FilterBuilder();
    this.filter.addDisjunction();
    this._resetMergeState();
  }

  addConjunction() {
    if (!this.filter) this.filter = new FilterBuilder();
    this.filter.addConjunction();
    this._resetMergeState();
  }

  flushMatches() {
    this.filter = null;
    this._resetMergeState();
  }

  _resetMergeState() {
    this.index = -1;
    this.currentKey = null;
    this.direction = null;
    this.realtimeSeekBound = null;
    this._resetCandidates();
  }

  queryUnique(fieldName) {
    const seen = new Set();
    const results = [];
    for (const r of this.readers) {
      for (const v of r.queryUnique(fieldName)) {
        const key = v.toString('base64');
        if (!seen.has(key)) { seen.add(key); results.push(v); }
      }
    }
    return results;
  }

  enumerateFields() {
    const fields = new Set();
    for (const r of this.readers) {
      for (const f of r.enumerateFields()) fields.add(f);
    }
    return fields;
  }

  listBoots() {
    const bootMap = new Map();

    for (const r of this.readers) {
      const bootId = uuidToString(r.header.tail_entry_boot_id);
      const firstSeq = r.header.head_entry_seqnum;
      const lastSeq = r.header.tail_entry_seqnum;
      const firstTime = r.header.head_entry_realtime;
      const lastTime = r.header.tail_entry_realtime;

      if (bootMap.has(bootId)) {
        const e = bootMap.get(bootId);
        if (firstSeq < e.firstSeq) e.firstSeq = firstSeq;
        if (firstTime < e.firstTime) e.firstTime = firstTime;
        if (lastSeq > e.lastSeq) e.lastSeq = lastSeq;
        if (lastTime > e.lastTime) e.lastTime = lastTime;
      } else {
        bootMap.set(bootId, { firstSeq, lastSeq, firstTime, lastTime });
      }
    }

    const boots = Array.from(bootMap.entries())
      .map(([bootId, t]) => ({ bootId, ...t }))
      .sort((a, b) => Number(a.firstTime - b.firstTime));

    const base = 1 - boots.length;
    return boots.map((b, i) => ({
      index: base + i,
      boot_id: b.bootId,
      first_entry: Number(b.firstTime),
      last_entry: Number(b.lastTime),
    }));
  }
}

function collectJournalFiles(path) {
  const files = [];
  const entries = readdirSync(path, { withFileTypes: true });

  for (const entry of entries) {
    const fullPath = join(path, entry.name);
    if (isRegularFile(fullPath) && isJournalFileName(entry.name)) {
      files.push(fullPath);
    }
  }

  for (const entry of entries) {
    if (!isJournalSubdirName(entry.name)) continue;
    const childPath = join(path, entry.name);
    if (!isDirectory(childPath)) continue;
    for (const child of readDirEntries(childPath)) {
      const childFile = join(childPath, child.name);
      if (isRegularFile(childFile) && isJournalFileName(child.name)) {
        files.push(childFile);
      }
    }
  }

  return files.sort();
}

function isRegularFile(path) {
  try {
    return statSync(path).isFile();
  } catch {
    return false;
  }
}

function isDirectory(path) {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
}

function readDirEntries(path) {
  try {
    return readdirSync(path, { withFileTypes: true });
  } catch {
    return [];
  }
}

function isJournalSubdirName(name) {
  if (name.includes('.')) return false;
  return id128StringValid(name);
}

function id128StringValid(s) {
  if (s.length === 32) return /^[0-9a-fA-F]{32}$/.test(s);
  if (s.length === 36) return /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/.test(s);
  return false;
}

function buildBootNewest(readers) {
  const newest = new Map();
  for (const reader of readers) {
    const bootId = Buffer.from(reader.header.tail_entry_boot_id);
    if (bootId.every(byte => byte === 0)) continue;
    const key = bufferKey(bootId);
    const current = newest.get(key);
    if (!current || reader.header.tail_entry_monotonic > current.monotonic) {
      newest.set(key, {
        machineId: Buffer.from(reader.header.machine_id),
        monotonic: reader.header.tail_entry_monotonic,
        realtime: reader.header.tail_entry_realtime,
      });
    }
  }
  return newest;
}

function cmpBigInt(a, b) {
  if (a < b) return -1;
  if (a > b) return 1;
  return 0;
}

function bufferEqual(a, b) {
  return Buffer.compare(a, b) === 0;
}

function bufferKey(buf) {
  return Buffer.from(buf).toString('hex');
}
